File: src/test/java/com/example/rpn/RpnTest.java
package com.example.rpn;

import org.junit.Test;
import static org.junit.Assert.*;


public class RpnTest
{
  private Calculator calculator = new RecursiveCalculator(AlgebraicOperators.OPERATORS);

  @Test
  public void testOperators()
  {
    assertEquals( 5.0, calculator.calculate("2 3 +").getRight(), 0.01);
    assertEquals(-1.0, calculator.calculate("2 3 -").getRight(), 0.01);
    assertEquals( 6.0, calculator.calculate("2 3 *").getRight(), 0.01);
    assertEquals(0.66, calculator.calculate("2 3 /").getRight(), 0.01);
  }

  @Test
  public void testExpressions()
  {
    assertEquals(-4.0, calculator.calculate("10 4 3 + 2 * -").getRight(), 0.01);
  }

  @Test
  public void testUnknownOperator()
  {
    assertTrue(calculator.calculate("1 2 ?").isLeft());
  }

  @Test
  public void testNotEnoughOperators()
  {
    assertTrue(calculator.calculate("1 2").isLeft());
  }

  @Test
  public void testNotEnoughArguments()
  {
    assertTrue(calculator.calculate("1 +").isLeft());
  }
}

File: src/test/scala/com/example/rpn/RpnTest.scala
package com.example.rpn

import org.junit.Test
import org.junit.Assert._


class RpnTest
{
  val calculator = Calculator(AlgebraicOperators.OPERATORS)

  @Test def testOperators =
  {
    assertEquals( 5.0, calculator.calculate("2 3 +").right.get, 0.01)
    assertEquals(-1.0, calculator.calculate("2 3 -").right.get, 0.01)
    assertEquals( 6.0, calculator.calculate("2 3 *").right.get, 0.01)
    assertEquals(0.66, calculator.calculate("2 3 /").right.get, 0.01)
  }

  @Test def testExpressions =
  {
    assertEquals(-4.0, calculator.calculate("10 4 3 + 2 * -").right.get, 0.01)
  }

  @Test def testUnknownOperator =
  {
    assertTrue(calculator.calculate("1 2 ?").isLeft)
  }

  @Test def testNotEnoughOperators =
  {
    assertTrue(calculator.calculate("1 2").isLeft)
  }

  @Test def testNotEnoughArguments =
  {
    assertTrue(calculator.calculate("1 +").isLeft)
  }
}
